core/matrix/sparsity.cpp

```diff
@@ -0,0 +1,195 @@
+                                                LinOp *x) const
+{
+    using Dense = Dense<ValueType>;
+    this->get_executor()->run(sparsity::make_advanced_spmv(
+        as<Dense>(alpha), this, as<Dense>(b), as<Dense>(beta), as<Dense>(x)));
+}
+
+
+template <typename ValueType, typename IndexType>
+void Sparsity<ValueType, IndexType>::read(const mat_data &data)
+{
+    size_type nnz = 0;
+    for (const auto &elem : data.nonzeros) {
+        nnz += (elem.value != zero<ValueType>());
+    }
+    auto tmp =
+        Sparsity::create(this->get_executor()->get_master(), data.size, nnz);
+    size_type ind = 0;
+    size_type cur_ptr = 0;
+    tmp->get_row_ptrs()[0] = cur_ptr;
+    for (size_type row = 0; row < data.size[0]; ++row) {
+        for (; ind < data.nonzeros.size(); ++ind) {
+            if (data.nonzeros[ind].row > row) {
+                break;
+            }
+            auto val = data.nonzeros[ind].value;
+            if (val != zero<ValueType>()) {
+                tmp->get_col_idxs()[cur_ptr] = data.nonzeros[ind].column;
+                ++cur_ptr;
+            }
+        }
+        tmp->get_row_ptrs()[row + 1] = cur_ptr;
+    }
+    tmp->move_to(this);
+}
+
+
+template <typename ValueType, typename IndexType>
+void Sparsity<ValueType, IndexType>::write(mat_data &data) const
+{
+    std::unique_ptr<const LinOp> op{};
+    const Sparsity *tmp{};
+    if (this->get_executor()->get_master() != this->get_executor()) {
+        op = this->clone(this->get_executor()->get_master());
+        tmp = static_cast<const Sparsity *>(op.get());
+    } else {
+        tmp = this;
+    }
+
+    data = {tmp->get_size(), {}};
+
+    for (size_type row = 0; row < tmp->get_size()[0]; ++row) {
+        const auto start = tmp->row_ptrs_.get_const_data()[row];
+        const auto end = tmp->row_ptrs_.get_const_data()[row + 1];
+        for (auto i = start; i < end; ++i) {
+            const auto col = tmp->col_idxs_.get_const_data()[i];
+            const auto val = one<ValueType>();
+            data.nonzeros.emplace_back(row, col, val);
+        }
+    }
+}
+
+
+template <typename ValueType, typename IndexType>
+std::unique_ptr<LinOp> Sparsity<ValueType, IndexType>::transpose() const
+{
+    auto exec = this->get_executor();
+    auto trans_cpy = Sparsity::create(exec, gko::transpose(this->get_size()),
+                                      this->get_num_nonzeros());
+
+    exec->run(sparsity::make_transpose(trans_cpy.get(), this));
+    return std::move(trans_cpy);
+}
+
+
+template <typename ValueType, typename IndexType>
+std::unique_ptr<LinOp> Sparsity<ValueType, IndexType>::conj_transpose() const
+    GKO_NOT_IMPLEMENTED;
+
+
+template <typename ValueType, typename IndexType>
+std::unique_ptr<Sparsity<ValueType, IndexType>>
+Sparsity<ValueType, IndexType>::to_adjacency_matrix() const
+{
+    auto exec = this->get_executor();
+    auto adj_mat = Sparsity::create(
+        exec, this->get_size(), this->get_num_nonzeros() - this->get_size()[0]);
```

Comment:
The matrix might have zero in diagonal, so You need a kernel to check the number of diagonal elements.

Reply:
The adjacency matrix (atleast if the graph it is representing has no loops) does not have anything except zeros on the diagonal. It is true that in the general case, there could be non-zeros on the diagonal, but I think when you convert a matrix to an adjacency matrix that is not possible(?).

Reply:
Yes, the adjacency matrix does not have diagonal elements.
Maybe I miss something.
I think the original matrix should have some zero in the diagonal.
The following 2 x 2 matrix has 2 element
```
1 0
1 0
```
the adjacency matrix size is not `2 (nnz) - 2 (num_row) = 0` .
```
0 0
1 0
```

Reply:
Ah, I misunderstood you. I see what you mean. I will add a kernel to count the number of diagonal elements and use that here. Thanks.